### ISSUE TITLE
Refactor file I/O for optimization and Unicode support

### ISSUE THREAD

**Issue by CSBVision** [opencv/opencv#24491]
### Describe the feature and motivation

Hello everyone,

We recently came accross the following two, related issues.
The first one refers to the image I/O currently implemented inside the `imgcodecs` module. With particular focus on reading files and decoding images - for encoding and writing output files the situation is analogous - the current approach is:
* calling `imread(filename)`
* openining the file / creating the respective `FILE*` pointer.
* start reading the file header to resolve the required decoder by `findDecoder(filename);`
* use the right `ImageDecoder` instance to read the file

This current approach involves that the file is always opened twice as the created file pointer is not passed to the decoder. Instead, the decoders use the filename only and have to re-open the file. This is potentially suboptimal as file access is usually much slower than memory access, which is particularly relevant if you open a lot of images, for example if OpenCV is used in combination with deep learning trainings. However, we did not benchmark these speed differences and they are not our main intention for the following proposal.

Alternatively, OpenCV supports decoding the image from a preloaded buffer by:
* reading the file content into a preloaded `std::vector<string> buffer` independent of OpenCV code
* calling `imdecode(buffer)`
* `imdecode()` resolves the decoder type from the buffer by `findDecoder(buf_row)`
* thereafter, the image is decoded from the buffer without opening the file twice (at least according to our code inspection)

If the decoder does not support reading from a buffer, OpenCV writes the buffer to a temporary file. However, all decoders do support it, with the only exception of the gdcm decoder, as given by the respective `m_buf_supported` value. Still, it supports reading from a stream such that a temporary file is still not necessary, however this is not of main interest here.

Consequently, the multiple file system calls by `imread()` can be optimized away in most cases by reading the file buffer first and, thereafter, passing the buffer to `imdecode()`. Our main intention for proposing this change is to extend OpenCV's file handling abilities. Currently, OpenCV uses `fopen()` to access files. This does a fine job on Linux (presumably the same holds for any Unix), however on Windows, `fopen()` has limited Unicode support. It was discussed before to [extend OpenCV's API on Windows](https://github.com/opencv/opencv/issues/4292#issuecomment-601686965), whereas the current proposal is to extend the API to [support streams](https://github.com/opencv/opencv/issues/15675). However, we think by reading the buffers inside OpenCV using the following function, full Unicode support will be available on Windows, too:

```
std::vector<char> readFile(const std::string& filename)
{
    std::ifstream stream(std::filesystem::u8path(filename).c_str(), ios::binary | ios::ate);
    std::vector<char> buffer;
    if (stream.good())
    {
        std::streamsize size = stream.tellg();
        stream.seekg(0);
        buffer.resize(size);
        stream.read(&buffer[0], size);
    }
    else
    {
        CV_LOG_WARNING(NULL, "readFile('" << filename << "'): can't open/read file: check file path/integrity");
    }
    return buffer;
}

// now refactor imread() to read the file first and use imdecode() thereafter
```

We think that using this approach inside `loadsave.cpp` will be not too complicated and significantly improves the usablity because presumably most UTF-8 Windows-only issues would be already solved while not needing any API changes. Or did we miss anything important here?

Even if the stream API will be implemented in advance, it can easily combined with our proposal by creating the stream first and, thereafter, decode the image directly from the stream instead of the buffer. Besides, this function could be exported to be used wherever file system access is required in other modules, too, either based on streams or buffers, respectively.

What do you think about this proposal?

**Comment by asmorkalov** [comment]
@vpisarev please take a look.

**Comment by weypro** [comment]
I would like to raise another question: Is there any plan to support `std:: u8string` in c++20? `std::filesystem::u8path` is deprecated in c++20.

**Comment by CSBVision** [comment]
True, still this does not really seem to be an issue. With C++ 20, there might a UTF-8 string such that migrating all code to this should be possible, maybe typedefing `cv::String` accordingly will even work while not breaking anything at all. Maybe interesting reference at [Stackoverflow](https://stackoverflow.com/questions/54004000/why-is-stdfilesystemu8path-deprecated-in-c20) regarding this.

By the way: Another interesting question might be what's actually the fastest way to read the file into the buffer. The current proposal could still be suboptimal as it reads the stream size first and thereafter copies its content. Maybe using a stringstream instance
```
ostringstream ss;
ss << stream.rdbuf();
buffer = ss.str();
```
could be faster?
We don't have a strict preference here... our main intention was to support Unicode paths on Windows, too, without needing any API changes. We just noticed that the file handling could be suboptimal, too. Appreciating any comment regarding both, whether this proposal is useful and what's actually the most efficient way to implement it.

**Comment by vpisarev** [comment]
@CSBVision,
thank you! This looks like an interesting proposal :) I have a few questions (and also, frankly speaking, some concerns):

1. what, in general, is a problem of opening a file twice? I believe, some profiling needs to be done, but my impression is that it's a fairly cheap operation. In principle, we could add extra methods to codecs that will take `FILE*` instead of a file name, then we can simply use `rewind()` function to reset the pointer to the beginning before passing it to the found codec.
2. you say, the new solution reads the whole file into memory and then calls imdecode. It may cause 2 problems:
   1. it may increase the memory consumption significantly in the case of image formats with light or no compression like Png, Tiff, OpenExr.
   2. whether protection from huge images will still work? We now have some protection from memory overflow when reading huge images. One obvious case is something like "zip-bomb" when a tiny png file may expand to a huge image; but there can be cases when the file itself is huge and simply does not fit into memory. See `validateInputImageSize()`. What about the proposed solution?
3. maybe we can try this approach with *1) read the whole file, 2) use imdecode* in some special cases (e.g. when we have a local file of a moderate size, e.g. a few hundred megabytes or less), but I'd probably prefer to use memory-mapped files to reduce memory consumption. But then it's not a cross-platform solution.
4. Regarding acceleration of the part that finds a proper decoder. Some slight improvement is possible if we just put the most popular cases in front (e.g. jpeg, png, webp), see `ImageCodecInitializer`. Now the codecs are placed in alphabetical order, which is probably wrong. But any solution with mutable storage may cause serious problems with reenterability. I would prefer to keep the order static. Just like in the case with double file opening, we need to profile this part and make sure that this is a bottleneck. I'm pretty sure it's not, but I may be wrong.

**Comment by Kumataro** [comment]
My opnion is ...
- Unfortunally `u8path` will be deprecated at C++20. See https://en.cppreference.com/w/cpp/filesystem/path/u8path
- Disk-cache on storage and file-cache on OS can be hide double-file-open performance issue (for sniffing and decoding).
- Memory Mapped File (mmap for posix, createFile for windows) makes slightly portability problem.
- Reading whole images is heavy cost. For example TIFF file can be contain many images until ~2GB (or ~4GB).

**Comment by nigels-com** [comment]
u8path is indeed deprecated, but we've worked around it with compiler flags for both gcc and msvc.
It's all well and good for C++ boffins to change their mind but upath is going be a around for a good long time, if not forever.

**Comment by mshabunin** [comment]
> Even more, we think this problem still exists at the moment: The codecs are initialized by `getCodecs()`. If multiple threads do so in parallel, there is no guarantee that the first one fully initializes `static ImageCodecInitializer g_codecs;`

I believe this is not the case since C++11 - standard guarantees thread-safe initialization of static block-scope variables, e.g. https://stackoverflow.com/questions/23593800/static-variable-initalization-syncronizes-in-c11

**Comment by CSBVision** [comment]
> > Even more, we think this problem still exists at the moment: The codecs are initialized by `getCodecs()`. If multiple threads do so in parallel, there is no guarantee that the first one fully initializes `static ImageCodecInitializer g_codecs;`
> 
> I believe this is not the case since C++11 - standard guarantees thread-safe initialization of static block-scope variables, e.g. https://stackoverflow.com/questions/23593800/static-variable-initalization-syncronizes-in-c11

Excellent, thanks for point out! Didn't know that this changed with C++ 11, so we are save here. 👍

**Comment by mshabunin** [comment]
BTW, in order to optimize decoder selection process we can introduce new functions or modes which would allow user to choose decoder manually or automatically based on extension/mime type, something smilar to backend selection in VideoCapture.

**Comment by CSBVision** [comment]
Yes agreed, we also thought about that but didn't implemented it so far.
Summary:
1. Add an enum `ImageType` or something similar that defines `PNG`, `JPG`, etc. as well as a special value `NONE`.
2. Introduce a function to map a file extension to it's equivalent enum representation, e.g. `"png"` to `ImageType::PNG`, `"jpg"` to `ImageType::JPG` but also `"jpeg"` to `"ImageType::JPG"`. Of course, analogous implementation for all possible encoders/decoders.
3. Save the encoders/decoders inside a `map<ImageType, ...>` where the value type is either the decoder or encoder. Currently, these are shared_ptrs to non-const objects. As any encoding/decoding uses a "fresh" encoder/decoder instance (-> thread-safety), we even suggest to use const instances to create the non-const, actually used ones. This introduces an additional protection towards unintended modifications, resulting in potential race conditions during implementing new encoders/decoders in the future.
4. To decode the image, first try to find a matching decoder by the file extension. Only if this does not work, try all and maybe trace a warning for inconsistent file names (e.g. PNG files named as `.jpg`).

If this is fine for you - currently it's just a proposal, we did not want to introduce to many inferences - we are open to implement it. It's not too complicated and can be kept inside `loadsave.cpp` without breaking anything else. Maybe we can even reuse some functionality from the videoio module, don't know whether something similar already exists there? Please let us know whether we should implement this proposal, too.

**Comment by Kumataro** [comment]
(this is only idea, I'm sorry if my opinion is strange.)
Introducing `datastrm` proxy may be useful to fix this problem.
(It is similar to `bitstrm`).

- Implement `datastrm` proxy to support (A) opening UTF-8 file path(using `std::filesystem::u8path`) and (B) opening memory buffer, and to provide `datastrm-context` and read/write/seek callback functions. 
  - For POSIX, it can work with `mmap()` function.
  - For Windows, it can work with `createFileMapping()` function(Maybe).
  - For others, it can work with FILE* contexts.
- `loadsave.c` creates `datastrm` instance with UTF-8 path or memory address.
- `loadsave.c` read signatures from `datastrm`, and use it to detect suitable decoder. 
- `loadsave.c` provide `datastrm-context` and callback functions to detected decoders.
- detected decoders sets `datastrm-context` and callback functions to specific codec library(`libjpeg`, `libtiff`, ...).
- codec library read data stream with callback functions and `datastrm-context` similar like to read from file.

Advantages are ...

- The target file is opened only once at `datastrm` instance.
- There is no need to read the entire file on working memory. This is good for performance and cost.
- Even if some compilers drop `std::filesystem::u8path` support completely(maybe stdc++28 or higher) , this effects are limited only in `datastrm` code.
- Even if the codec library does not support UTF-8 path( build option or old version), it may work well.

Disadvantages are...
- We have to update decoders to use `datastrm-context` and  callback functions.

**Comment by CSBVision** [comment]
The idea sounds good, but is it really a feasible approach? We think from inspecting the current implementation, one major issue is that all third-party codec libraries don't expose a standardized API. All seem to support direct file access, while most support decoding from a binary raw buffer, i.e. `const char*`, `void*` or something similar (cf. `m_buf_supported` flag), too. Currently, gdcm seems to be the only decoder that does not support this, and requires a filename (in particular, `imdecode` writes a temporary file in this case that is only used to decode the image). Still, streams would be supported here.

To us this means that without controlling the APIs of the actual encoder/decoder libraries, a proxy class/struct would have to contain *all* available information (filename, buffer, stream, maybe more?), including the possibility that some are not available (e.g. encoding a buffer alone leaves filename empty). Then, passing the proxy instance to the decoder is possible, the decoder picks the best match (e.g., buffer supported and available? -> then we take this one). On the other hand, it might create a temporary filename (e.g., nothing else could be supported by the decoder) and writes this into the proxy. Unavoidably, this moves some of the now centralized logic (buffer supported? temporary file?) to the individual encoder/decoder implementations.

Alternatively, it could be reasonable to extend the decoder APIs inside OpenCV with a flag `m_stream_supported` and introducing read/write functions operating on the stream, cf. #15675. But this could be addressed independently of our proposed change... we could simply implement the stream API in all decoders - which does not interfere with `loadsave.cpp` - and thereafter benchmark whether the buffer or the stream variant is faster by default, i.e. which one should be used with higher priority if both are supported.

**Comment by Kumataro** [comment]
Yes, so I said "this is only idea". Further consideration is needed to achieve this.

But the approach of reading all compressed data streams(JPEG/TIFF/WEBP/...) at once is very difficult/hard for some low resource / embedded environments, like SBCs (Single board Computers) which include less than 4GB and cannot extend memory.

- Rasberry Pi 4 has 1GB memory model.
- Rasberry Pi 5 has 2GB memory model.

If data, which can be decoded previous OpenCV, can no longer be decoded with latest OpenCV, this will be recognized as a "bug" or "degraded" by the user.
(Even if there is a way to increase physical memory or SWAP memory).

This is my concern.

**Comment by CSBVision** [comment]
Yes, we can understand these concerns, but are they really that justified?

First, it is still possible to move to a stream-based decoding for most decoders such that even such an issue could be resolved, too, however requires to extend the decoder APIs and, as a matter of fact, cannot be kept in `loadsave.cpp` alone (as discussed before). Second, multi-page reads remain unchanged (we only implemented unicode support here).

Focusing on singe-page images, the decoded image itself usually is larger than the file, for compressed encoders often significantly larger (for uncompressed/raw encoders, maybe the file is slightly larger than the decoded image in memory from container overhead, metadata or something else, but that should be negligible). Next, the decoded image itself has to fit in memory. If not, it makes no sense at all to decode it and it's the user's responsibility of load only images that do (accidentally opening too large files should result in an empty result or error, this might occur during reading the whole file in memory as well). So where is the realistic scenario where the image does fit in memory and can be processed, **but** the content of the file from which it was decoded would take too much memory?

First, the overhead only exists during decoding. As soon as the image is fully decoded, the buffer is released, and the overhead is gone. The image has to be rather big (>100 MB), the system memory rather low and there is no virtual/swap memory. Then, there could be a theoretical setting where the overhead from reading the whole stream at once might become problematic. We might be wrong, but are these scenarios really realistic somehow? Even copying the image is not possible (if the file does not fit in memory, a copy of the image won't do either), processing, encoding to a compressed format etc. easily exceed the system memory as well. Not to mention any processing speeds of low-resource devices on huge images. Long story short: To us these settings seem like pathological examples but aren't even close to a realistic use case.

One the other hand, RAM is so much faster than file access. We have to read the file, so reading it only once, read everything at all without sequential reads presumably is the fastest way we can do. So how to proceed? Should we create a PR such that you could review our implementation? Or shall we implement a different data structure for the decoders before and create a PR after this is completed, too? Looking forward to your comments 👍

**Comment by Kumataro** [comment]
I think the debate about whether or not to introduce **breaking changes** should be considered carefully.
Or I suggest to limit side-effects, to add `IMREAD_FORCE_READ_AT_ONCE` option to support reading all compressed image data with utf-8 path. This solutioin saves both back-compatibility(memory usage) and accessing utf-8 path.

> (for uncompressed/raw encoders, maybe the file is slightly larger than the decoded image in memory from container overhead, metadata or something else, but that should be negligible)

I'm sorry but I disagree with it. Realistic data contradicts this.

Generally decoder and codec use 3 kind memories to decode compress image to RAW image.
- Compressed file (?)
- Working memory (~2GB)
- RAW image(~2GB)

RAW image limits is 2GB usually by filesystem or memory on OS. 
Sometimes working memory requests the same or 1/n size of RAM image.

How about the size of compress file ?

Usually natual picture images can be compressed lossy with about 1/5 ~ 1/10.
If we use lossless compressed more smaller.

But some image file format are data container which can archive more information.
PNG, JPEG, TIFF format can be store comment data stream.
TIFF, JPEG XL format can be store multiplages.
And some file formats can be include thumbnail images.

This means user application can be stores appendix information until file format limitation.
This information is not guaranteed to be smaller than the compressed image data.

And Is it the user's responsibility whether the data they enter is processed ?
If so, we must clarify the terms on the document to keep it.
They are not one of broken files and too large files.

I repeat that it is not desirable to be unable to open files or to have increased memory usage simply by upgrading the OpenCV. User will think that is OpenCV library responsibility.

---
In abstract terms:

- 95% of our users won't notice the change.
- 4% of our users will be happy. (supporting utf-8 path)
- 1% of our users will be unhappy. (increase memory usage)

I believe we have to consider about whether to accept this fix or consider other workarounds.

**Comment by nigels-com** [comment]
Chiming in from the point of view of our application of OpenCV.
We _care_ about utf-8 paths on both Linux and Windows.
A minor API break is manageable for us.
Memory use is not a significant consideration.
Reading once rather than twice is nice to have.  (1/100 of our desire for utf-8 paths on Windows)

**Comment by Kumataro** [comment]
Thank you and it is OK, this is last message from me about this topic.  It is very unfortunate that sufficient consideration has not been given to users in low memory environments(like embedded and edge computing).

**Comment by vpisarev** [comment]
@CSBVision,

ok, let me provide at least intermediate summary based on our discussion.

1. using u8path to provide proper argument for fopen/_wfopen is probably fine, despite deprecation.
2. using imdecode to prevent double-opening and to reduce the number of places to insert u8path is no-go. It will create more problems than it will solve. We won't accept it. There are alternative methods to avoid double-opening and to add u8path, which can be implemented rather easily, maybe with just a little more code.
3. using complex tricks and introduction of an extra API to speedup codec selection is no-go. Just reorder codecs during imgcodecs initialization and the problem will mostly be solved. If there is any problem at all. In any case, such a patch, if it's submitted, _must_ come together with a detailed study why this needs to be accelerated and what is the result of this acceleration or 'acceleration'. OpenCV is already quite complex library and yet we constantly try to keep the API simple and also to keep it robust under in a wide range of scenarios. Let's not over-design things. Any extra API is a potential big headache for us for many years. I will try to stop all such efforts as much as I can.

**Comment by vpisarev** [comment]
Some extra words about the item 3. In imgcodecs we already have `haveImageReader()` function that basically runs the codec selection part alone, without further decoding. There can be some experiment where an std::vector of full file names for a big collection of images, e.g. 100,000 images is created and then we do, say, 1,000,000 iterations and on each iteration we select a random index from the vector and do `printf("%d. have decoder for '%s': %d\n", iter, filename[idx].c_str(), (int)haveImageReader(filename[idx]);`. It would be interesting to see how much time does it take to perform 1,000,000 of fopen's together with the codec selection, and compare it to the training time.

**Comment by CSBVision** [comment]
Thanks for your comment and let's clarify: Our main intention was neither to introduce any breaking changes - other people proposed it in this discussion, we always tried to keep it simple / only in `loadsave.cpp` - nor to mix around APIs, or even change how main parts of OpenCV work. The starting point was unicode support, which is simply not available, as stated already in the opening comment

> Our main intention for proposing this change is to extend OpenCV's file handling abilities.

as we thought this should be an easier solution to already existing issues. We thought it should reasonable to directly read the file beforehand, and avoid all the code duplicates between `imread` and `imdecode` as well as the `findDecoder`. But we are fine with other opinions. If you want to keep your implementation - fine for us, we can use ours by wrapping `imdecode`. That's why we did not a create a PR so far, but instead wanted to discuss the direction first. 👍 

Alternatively, we think unicode support should be possible by using `u8path` before `fopen`, even without anything else. Therefore, we propose to redesign our patch with the following features, and we strip out everything else for the moment:
1. Introduce a custom `fopen` function that uses `u8path` internally first to resolve all unicode issues before `fopen`, if C++ 17 is available, and directly calls `fopen` otherwise (clearly, without resolving the unicode issues).
2. Introduce a `maxSignatureLength()` function that queries the signature length from all decoders and returns the maximum one. This should be called by both `findDecoder` overloads and cache the result in a `static const size maxlen` variable. Currently, this is queried from all decoders during each decoding over and over again, which can simply be cached.
3. Reorder the codecs to move popular ones in front, maybe `[png, jpg, webp, tif, etc.]`  is at least reasonable based on [statistics](https://w3techs.com/technologies/overview/image_format).

All follow-up optimizations to resolve the codec - as stated above, this was not our main intention, we only observed this while looking through the code - can be addressed independently in a different patch (use a different data structure? resolve from file extension? cache the last success in an atomic variable?). Maybe we could start by writing a small benchmark app, similar to your suggestion, and try to find out whether that's worth the effort. Still, even if the benefit might be small, it makes not too much sense to query rare formats first just because they're listed first in an alphabetical order. Thus, we kept this proposal on the list.

What do you think about this refactoring?

**Comment by CSBVision** [comment]
Alright, we will prepare a respective PR. Still, the proposal with the `fopen` wrapper function leads to another issue: We need to replace all `fopen` calls with this new function. Still, if the decoder internally calls `fopen`, this is not possible. So we need to call `u8path` before and pass the resolved filename. Therefore, we propose two functions in `filesystem.hpp`: The first one wraps `u8path` (if C++ 17 is available) and returns a string, whereas the other calls the first and, thereafter, `fopen`.
Is this fine, too, or should we resolve the the filename only in `loadsave.cpp`, and keep `filesystem.hpp` unchanged?
